include/envoy/server/listener_manager.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "envoy/json/json_object.h"
+#include "envoy/network/filter.h"
+#include "envoy/network/listen_socket.h"
+#include "envoy/ssl/context.h"
+
+namespace Envoy {
+namespace Server {
+
+/**
+ * Factory for creating listen sockets.
+ */
+class ListenSocketFactory {
+public:
+  virtual ~ListenSocketFactory() {}
+
+  /**
+   * Creates a socket.
+   * @param address supplies the socket's address.
+   * @param bind_to_port supplies whether to actually bind the socket.
+   * @return Network::ListenSocketPtr an initialized and potentially bound socket.
+   */
+  virtual Network::ListenSocketPtr create(Network::Address::InstanceConstSharedPtr address,
+                                          bool bind_to_port) PURE;
+};
+
+/**
+ * A configuration for an individual listener.
+ */
+class Listener {
+public:
+  virtual ~Listener() {}
+
+  /**
+   * @return Network::FilterChainFactory& the factory for setting up the filter chain on a new
+   *         connection.
+   */
+  virtual Network::FilterChainFactory& filterChainFactory() PURE;
+
+  /**
+   * @return Network::Address::InstanceConstSharedPtr the address.
+   */
+  virtual Network::Address::InstanceConstSharedPtr address() PURE;
+
+  /**
+   * @return Network::ListenSocket& the actul listen socket. The address of this socket may be
+   *         different from address() if for example the configured address binds to port 0.
+   */
+  virtual Network::ListenSocket& socket() PURE;
+
+  /**
+   * @return Ssl::ServerContext* the SSL context
+   */
+  virtual Ssl::ServerContext* sslContext() PURE;
+
+  /**
+   * @return bool whether to use the PROXY Protocol V1
+   * (http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt)
+   */
+  virtual bool useProxyProto() PURE;
+
+  /**
+   * @return bool specifies whether the listener should actually listen on the port.
+   *         A listener that doesn't listen on a port can only receive connections
+   *         redirected from other listeners.
+   */
+  virtual bool bindToPort() PURE;
+
+  /**
+   * @return bool if a connection was redirected to this listener address using iptables,
+   *         allow the listener to hand it off to the listener associated to the original address
+   */
+  virtual bool useOriginalDst() PURE;
```

Comment:
Sure. FYI, we are ready to use the v2 protos now in the Envoy main repo, all the import stuff is setup.